airbyte-webapp/src/components/connectorBuilder/Builder/StreamReferenceField.tsx

```diff
@@ -0,0 +1,58 @@
+import { useField } from "formik";
+import { useMemo } from "react";
+import { FormattedMessage } from "react-intl";
+
+import { ControlLabels } from "components/LabeledControl";
+import { DropDown } from "components/ui/DropDown";
+import { Text } from "components/ui/Text";
+
+import { BuilderStream } from "../types";
+import styles from "./BuilderField.module.scss";
+
+interface StreamReferenceFieldProps {
+  // path to the location in the Connector Manifest schema which should be set by this component
+  path: string;
+  label: string;
+  tooltip?: string;
+  optional?: boolean;
+  currentStreamIndex: number;
+}
+
+export const StreamReferenceField: React.FC<StreamReferenceFieldProps> = ({
+  path,
+  label,
+  tooltip,
+  optional,
+  currentStreamIndex,
+  ...props
+}) => {
+  const [streams] = useField<BuilderStream[]>("streams");
+  const [field, meta, helpers] = useField(path);
+  const hasError = !!meta.error && meta.touched;
+
+  const options = useMemo(() => {
+    return streams.value
+      .filter((_value, index) => index !== currentStreamIndex)
```

Comment:
To solve the circular dependency case that you called out, maybe we could expand this filter logic to also exclude streams that are descendants of the current stream.

I think this would require keeping track of the whole tree of substream parents in a state somewhere and will definitely add some complexity, so we can do that in a separate PR. Want to create an issue for that?

Reply:
Yeah, sounds good, gonna create an issue for that.